### PR TITLE
Allow threads to be shown for hints

### PR DESCRIPTION
**Pull request by digit01Wave**
#974 

When we have hints, allow users to unlock help threads

### Before unlock:
image
image

### After unlock
image

### When try to access thread that has not been unlocked
image

### CODE REVIEW

File: ServerCore/Pages/Threads/PuzzleThread.cshtml.cs [hunk 1]
```diff
@@ -74,7 +74,27 @@ public async Task<IActionResult> OnGetAsync(int? puzzleId, int? teamId, int? pla
...
             }
 
-            Team team = null;
+            // If we are showing hints, only allow the thread to be seen if the user has unlocked it
+            if (!Event.HideHints)
+            {
+                string returnUrl = HttpContext.Request.Path.HasValue ? HttpContext.Request.Path.Value : null;
+                if (Event.DefaultCostForHelpThread < 0)
+                {
+                    return RedirectToPage("/ErrorPage", new { ErrorMessage = "Help threads are not allowed for this event" });
```
Comment (tabascq) [inline ServerCore/Pages/Threads/PuzzleThread.cshtml.cs:83]: Is this something that can happen without the player customizing the URL? If not, then I think `return Forbid();` is enough.

Comment (tabascq) [reply ServerCore/Pages/Threads/PuzzleThread.cshtml.cs:83]: Same goes for the redirect on line 93.

Comment (digit01Wave) [reply ServerCore/Pages/Threads/PuzzleThread.cshtml.cs]: done

File: ServerCore/Pages/Teams/Hints.cshtml [hunk 2]
```diff
@@ -90,16 +91,25 @@ You have <b>
...
                     @if (hint.IsUnlocked)
                     {
-                        @hint.Hint.Content
+                        if (hintViewState != null)
+                        {
+                            @hint.Content
+                        }
+                        else if (puzzleThreadViewState != null)
+                        {
+                            <a asp-Page="/Threads/PuzzleThread" asp-route-puzzleid="@Model.PuzzleID" asp-route-teamId="@Model.Team?.ID" asp-route-playerId="@Model.LoggedInUser.ID">Use this link to message author!</a>
```
Comment (tabascq) [inline ServerCore/Pages/Teams/Hints.cshtml:111]: If there's room, "Use this link to message **the** author!" would be clearer.

Comment (digit01Wave) [reply ServerCore/Pages/Teams/Hints.cshtml]: done!

**Review by tabascq — APPROVED** [review]

**Pull request merged**